Plugins/src/main/java/tw/waterball/judgegirl/plugins/impl/cqi/AbstractJudgeGirlCodeQualityInspectionPlugin.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2020 Johnny850807 (Waterball) 潘冠辰
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package tw.waterball.judgegirl.plugins.impl.cqi;
+
+import tw.waterball.judgegirl.plugins.api.AbstractJudgeGirlPlugin;
+import tw.waterball.judgegirl.plugins.api.codeinspection.CodeQualityInspectionReport;
+import tw.waterball.judgegirl.plugins.api.codeinspection.JudgeGirlCodeQualityInspectionPlugin;
+import tw.waterball.judgegirl.cqi.cyclomatic.CyclomaticComplexityReport;
+
+import java.nio.file.Path;
+import java.util.Map;
+
+/**
+ * @author - johnny850807@gmail.com (Waterball)
+ */
+public abstract class AbstractJudgeGirlCodeQualityInspectionPlugin extends AbstractJudgeGirlPlugin
+        implements JudgeGirlCodeQualityInspectionPlugin {
+    protected Map<String, String> parameters;
+
+    public AbstractJudgeGirlCodeQualityInspectionPlugin(Map<String, String> parameters) {
+        super(parameters);
+        this.parameters = parameters;
+    }
+
+    @Override
+    public CodeQualityInspectionReport performAtSourceRoot(Path sourceRootPath) {
+        return new CodeQualityInspectionReport(calcCyclomaticComplexity(sourceRootPath));
+    }
+
+    protected abstract CyclomaticComplexityReport calcCyclomaticComplexity(Path sourceRootPath);

```

Comment:
This is good, [template method](https://sourcemaking.com/design_patterns/template_method)